ENH: Add the alternating dual update method

sbanert opened the pull request:
This pull request adds a new (nonsmooth) solver, which is, according to the authors of the corresponding paper, especially good in doing tomography problems on a graphics card.

It contains the changes of #1166, and the corresponding test uses the functionality provided by #1166 as requested by Jonas in the discussion there.

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+            tmp_ran = tmp_rans[ops[i].range]
+            proxs[i](arg, out=tmp_ran)
+            x -= 1.0 / stepsize * ops[i].adjoint(tmp_ran - duals[i])
+            duals[i] = tmp_ran.copy()
```
duals[i].assign(tmp_ran)

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+        else:
+            rng = range(len(ops))
+
+        for i in rng:
```
Both loops use `i` as iteration variable, this is confusing. Prefer using different names, e.g. i and j

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+        for i in rng:
+            step = stepsize / majs[i]
+            arg = duals[i] + step * ops[i](x)
+            tmp_ran = tmp_rans[ops[i].range]
```
This whole code creates copies all over the place, I think it would be better to simply remove this temporary for now (alternatively fix all the copies, but that would be some work and I'm not sure how much it will help atm).

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+allows an acceleration by permitting efficient calculations on a graphics card.
+"""
+
+# Imports for common Python 2/3 codebase
```
Yes we removed those comments.

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Alternating dual (AD) upgrade algorithm studied by McGaffin and Fessler.
```
below it says "update" not "upgrade"

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+
+
+def adupdates(funcs, ops, x, stepsize, majs, niter, random=False,
+              callback=None, callback_loop='outer'):
```
I suggest that we use the same signature (where possible) as for the Douglas-Rachford method, i.e. `g` for functionals that are composed, `L` for the operators, etc.
See also #1197

sbanert replied:
I am currently doing this and I noticed that there is the parameter kwargs, which is not documented in the docstring of the Douglas–Rachford method. In the iterative methods, this parameter is not present at all. Which variant shall I implement here?

kohr-h replied:
We don't really have a rule on that, but if you feel that the signature gets a bit too long with all the optional parameters, stuff the less common, less important or more advanced ones into `kwargs`. In this case, I'd say it's fine without `kwargs` since you have only one optional parameter `random` (I don't count `callback`s).

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+    ----------
+    [MF2015] McGaffin, M G, and Fessler, J A. *Alternating dual updates
+    algorithm for X-ray CT reconstruction on the GPU*. IEEE Transactions
+    on Computational Imaging, 1.3 (2015), pp 186--199.
```
Reference looks perfect. If you want to use a citation with link, which would be something for the short summary before `Parameters`, you can use a link like
```
`[MF2015] <URL for DOI, arXiv or whatever>`_
```
This is at least my current favorite way of doing it, because it provides a clickable link *and* a text reference in the `References` section.

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,177 @@
+            if callback is not None and callback_loop == 'inner':
+                callback(x)
+        if callback is not None and callback_loop == 'outer':
+            callback(x)
```
This is a pretty complex optimization method. I'm sure it's a good one.
What I can recommend is to make a "dead simple" implementation of the same optimizer that does not do any fancy stuff like caching, in-place operation of optimized arithmetic. Just with as little code as possible. That's really good for debugging. See for example [here](https://github.com/odlgroup/odl/blob/87e04941f4337de190654897691327f8964c0f03/odl/solvers/nonsmooth/admm.py#L149-L163).

sbanert commented:
Should be ready from my side if Travis does not complain.

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    Parameters
+    ----------
+    g : sequence of `Functional`s
+        All functions need to provide a `convex_conj` with a
```
dead link here, you need to write `Functional.convex_conj`

sbanert replied:
I tried this by building the documentation. `Functional.convex_conj` is a dead link, too.

kohr-h replied:
I've never really understood the rules as to when you can use the name without full path. @adler-j can you nail it down?

sbanert replied:
Maybe it can’t find a destination. Links would have to go to entries of the attributes table in this file: https://odlgroup.github.io/odl/generated/odl.solvers.functional.functional.Functional.html

Actually, it would be nice to have the complete docstring of Functional.proximal and Functional.convex_conj printed somewhere in the docs.

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    ----------
+    g : sequence of `Functional`s
+        All functions need to provide a `convex_conj` with a
+        `proximal` factory.
```
Same as above, `Functional.proximal`

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+        All functions need to provide a `convex_conj` with a
+        `proximal` factory.
+    L : sequence of `Operator`s
+        Needs to provide as many elements as `g`.
```
This sentence could use some clarification. What is provided? Typically we say that the length of the sequence should be something

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    majs : sequence of majorizers
+        Parameters determining the stepsizes for the inner iterations. Must be
+        matched with the norms of `L`, and convergence is guaranteed if
+        `majs` are large enough. See the Notes section for details.
```
majs is a dead link, use double \`\` for non-link

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+        The stepsize for the outer (proximal point) iteration. The theory
+        guarantees convergence for any positive real number, but the
+        performance might depend on the choice of a good stepsize.
+    majs : sequence of majorizers
```
majorizers is still not very obvious. What does this mean? If it is functions, just write "sequence of callables"

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+        Number of (outer) iterations.
+    random : bool, optional
+        If `True`, the order of the dual upgdates is chosen randomly,
+        otherwise the order provided by the lists `g`, `L` and `majs`
```
dead links

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+        Function called with the current iterate after each iteration.
+    callback_loop : {'inner', 'outer'}, optional
+       If 'inner', the `callback` function is called after each inner
+       iteration, i.e., after each dual update. If 'outer', the `callback`
```
dead link

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    :math:`x`. This corresponds to the formula
+
+    .. math::
+         \tilde x_{n+1} = x_n - \frac{1}{\mu} \sum_{i=1}^m L_i^* v_i.
```
What does tilde mean here?

kohr-h replied:
Remark here that applies to a lot of places in our docstrings:

If we use [raw string literals](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals) whenever there's math in it, we can ditch all those double backslash ugliness.

Here, some math won't work (e.g. `\frac`), so the simplest fix is to just prepend the whole docstring with an `r`.

However, we should explore what happens with unicode stuff, in cases where names require non-ASCII characters.

sbanert replied:
Unicode seems to work fine at least in my environment with Python 3.6.

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    """
+    # Check the lenghts of the lists (= number of dual variables)
+    length = len(g)
+    if not len(L) == length:
```
```python
if len(L) != length:
```

is the preferable style

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+    # Check if operators have a common domain
+    # (the space of the primal variable):
+    domain = L[0].domain
+    if any(domain != opi.domain for opi in L):
```
prefer swapping the order here to follow natural language, e..g. you would typically not say "if the expected domain is not equal to the domain", rather you would say "if the domain is not equal to the expected domain".

In code,

```python
if any(opi.domain != domain for opi in L):
```

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+
+    if not len(majs) == length:
+        raise ValueError('Number of `majs` {} and number of `g` {} do not'
+                         ' agree.'.format(len(majs), length))
```
We typically write the exception messages in lower case characters and without a dot, e.g.

```python
'number of `majs` {} and number of `g` {} do not agree'
```

See https://github.com/odlgroup/odl/pull/427 if you want a rationalle for this. Basically we follow the python core standard.

kohr-h replied:
Small addition to this: it's more helpful for the user if the expected behavior is explicitly stated, e.g.,
```
'`len(majs)` should be equal `len(g)`, but {} != {}'
```
The library could definitely use a refreshment in that department, but new code should act as a good example 😄

kohr-h replied:
"equal to" that is

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+
+import numpy as np
+
+__all__ = ('adupdates', 'adupdates_simple')
```
The function `adupdates_simple` should not be exported.

sbanert replied:
But if I do not export this function, how can I use it in tests or for debugging?

adler-j replied:
It is still available under the full path `odl.solvers.nonsmooth.adaupdates.adaupdates_simple`.

kohr-h replied:
Technically, the `__all__` thing is for defining explicitly what should be exposed to wildcard imports, i.e. `from odl.solvers.nonsmooth.adaupdates import *`. If it's undefined, the wildcard import just imports everything not starting with an underscore (roughly).
For us, defining stuff in `__all__` is also a mechanism to make things available on a higher level, since the `__init__.py` files on higher levels do the wildcard import for all submodules (with exceptions).

So it's always possible to import names by going all the way down to where the name is actually defined (= bound to a value).

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Test for the Fessler alternating dual updates solver."""
```
Since "Fessler" is not part of the ODL name, it should not be mentioned here either

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+"""Test for the Fessler alternating dual updates solver."""
+
+from __future__ import division
+# import pytest
```
Remove this line

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+from __future__ import division
+# import pytest
+import odl
+# import numpy as np
```
Remove this line

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+
+    We therefore solve the problem
+
+    min ||Ax - b||^2 + TV(x) s.t. x >= 0
```
I was fine until this point. Why should this have the same solution as above? What does TV even mean here?

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+    mat1op = odl.MatrixOperator(mat1)
+    mat2op = odl.MatrixOperator(mat2)
+    domain = mat1op.domain
+    tv1 = odl.MatrixOperator([[1.0, -1.0, 0.0, 0.0], [0.0, 0.0, 1.0, -1.0]])
```
break line here for better readability

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+    rhs2 = mat2op.range.element(mat2op(expected_solution))
+
+    # Create the functionals
+    fid1 = odl.solvers.functional.functional.FunctionalTranslation(
```
No need to use `odl.solvers.functional.functional.FunctionalTranslation`, simply use `.translated(...)` on the functional

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+
+    expected_solution = domain.element([1, 1, 1, 1])
+    # Create right-hand-sides of the equation
+    rhs1 = mat1op.range.element(mat1op(expected_solution))
```
`mat1op.range.element` should not be needed here

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+    fid2 = odl.solvers.functional.functional.FunctionalTranslation(
+        odl.solvers.functional.default_functionals.L2NormSquared(
+            mat2op.range), rhs2)
+    reg1 = odl.solvers.functional.default_functionals.L1Norm(tv1.range)
```
odl.solvers.L1Norm(tv1.range) should suffice

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,106 @@
+    x = tv1.domain.zero()
+    x_simple = tv1.domain.zero()
+
+    # Select stepsize one
```
This line is somewhat redundant, it follows from the variable name. Documentation should not document "obvious" stuff. In general, the rule is that documentation should be about "why" something is done. "What" is done should follow from the code.

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+# Copyright 2014-2017 The ODL contributors
```
I prefer one sweeping update for just that.

kohr-h commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,210 @@
+
+    The Alternating Dual (AD) updates method of McGaffin and Fessler `[MF2015]
+    <http://ieeexplore.ieee.org/document/7271047/>`_ is designed to solve an
+    optimization problem of the form
```
We need something here, yes. If the text ends like this
```
text text text::
    code code code
```
the colon will be rendered by Sphinx. If the text should not end with a colon (like here), place a whitespace before the double colon, like this:
```
text text text ::
    code code code
```

pep8speaks commented:
Checking updated PR...





 - In the file [`odl/test/solvers/nonsmooth/adupdates_test.py`](https://github.com/odlgroup/odl/blob/855eef362e6af3bce19d2a97b4ec22735172cc0b/odl/test/solvers/nonsmooth/adupdates_test.py), following are the PEP8 issues :

> [Line 92:80](https://github.com/odlgroup/odl/blob/855eef362e6af3bce19d2a97b4ec22735172cc0b/odl/test/solvers/nonsmooth/adupdates_test.py#L92): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

 - Complete extra results for this file :

>     adupdates_method_simple(x_simple, funcs, ops, stepsize, inner_stepsizes, niter)                                                                               ^---

sbanert commented:
I did everything except

* the dead links Functional.convex_conj and Functional.proximal
* upgrading to 2018.

adler-j commented on odl/solvers/nonsmooth/adupdates.py:
```diff
@@ -0,0 +1,217 @@
+
+import numpy as np
+
+__all__ = ('adupdates_method',)
```
Why is it called "method" btw, wouldn't `adaupdates` look nicer?

adler-j commented on odl/test/solvers/nonsmooth/adupdates_test.py:
```diff
@@ -0,0 +1,102 @@
+
+    for the matrix A, the r.h.s. b as above and the total variation TV, which
+    is given as the sum of the absoulute values of consecutive entries of the
+    solution. The solution of this problem is clearly x = (1, 1, 1, 1), since
```
Well this is assuming that we're using periodic boundary conditions for TV (which is not default in ODL), but I guess we can let that pass for now :O

sbanert replied:
Not really, this total variation takes three differences for a four-component vector. This is still not the standard behaviour, though, because in ODL the absolute value of the last element would be added to the functional (which was very irritating to most of my students).

sbanert commented:
I fixed the issues and corrected a bug. Please do *not* merge, though, because this is a rebase on the branch of #1283 (otherwise #1279 would have been triggered).

sbanert commented:
Now, I rebased on master, but I think I will have to rebase once more because Travis will probably fail.

kohr-h commented:
Seems so, but after that it should all be fine.

aringh commented:
The fix for Travis doc build is now in master, so if you rebase again all test should hopefully pass :-)

adler-j commented:
Please perform that rebase!

sbanert commented:
Succeeded.

adler-j commented:
Merged, well done!
